Add function to compute weak flux divergence

nilsdeppe opened the pull request:
## Proposed changes

Adds a function to compute the weak form flux divergence. The code is optimized to reduce allocations and operate on contiguous buffers, hence the 1d, 2d, and 3d cases being handled separately.

### Types of changes:

- [ ] Bugfix
- x ] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add function to compute weak flux divergence

kidder commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,191 @@
+    divergence_of_flux->initialize(fluxes.number_of_grid_points());
+  }
+
+  // Multiplies the flux be det_jac_time_inverse_jacobian.
```
be -> by

kidder commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,191 @@
+          const gsl::not_null<Variables<tmpl::list<Tags::div<FluxTags>...>>*>
+              result_buffer,
+          auto xi_or_eta_jacobian) noexcept {
+        using tag = tmpl::type_from<decltype(tag_v)>;
```
tag -> flux_tag

kidder commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,191 @@
+          const auto result_tensor_index =
+              result.get_tensor_index(result_storage_index);
+          const auto x_tensor_index = prepend(result_tensor_index, 0_st);
+          auto y_tensor_index = x_tensor_index;
```
is this really faster than another call to prepend?

nilsdeppe replied:
No idea, I found it easier to read but am happy to change to prepend :)

kidder commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,191 @@
+             result_storage_index < result.size(); ++result_storage_index) {
+          const auto result_tensor_index =
+              result.get_tensor_index(result_storage_index);
+          const auto x_tensor_index = prepend(result_tensor_index, 0_st);
```
x_flux_tensor_index or flux_x_tensor_index ? (and similarly for the y and z components...)

kidder replied:
ping on this? you can do it when you squash; I'm happy with the other changes

kidder commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,191 @@
+          auto tag_v,
+          const gsl::not_null<Variables<tmpl::list<Tags::div<FluxTags>...>>*>
+              result_buffer,
+          auto xi_or_eta_jacobian) noexcept {
```
xi_or_eta_jacobian -> logical_index_of_jacobian

kidder commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,191 @@
+      };
+
+  if constexpr (Dim == 1) {
+    (void)det_jac_times_inverse_jacobian;  // is identically 1.0 is 1d
```
is 1d -> in 1d

kidder commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,191 @@
+                     matrix.spacing(),  // rows of matrix including padding
+                     input,             // u
+                     matrix.columns(),  // rows of u
+                     add_to_result ? 1.0 : 0.0,  // multiplier for unused term
```
what is meant by unused term? aren't you using it if add_to_result is true?

nilsdeppe replied:
relic of an old, no longer relevant comment. I'll update

kidder commented on tests/Unit/NumericalAlgorithms/LinearOperators/Test_WeakDivergence.cpp:
```diff
@@ -0,0 +1,163 @@
+                            &local_approx, &mesh](auto tag_v) {
+    using tag = tmpl::type_from<decltype(tag_v)>;
+    if (mesh.quadrature(0) == Spectral::Quadrature::GaussLobatto) {
+      // Only the interior points are zero. Points closer at the boundary
```
Points on the boundary ?

nilsdeppe commented:
Okay, pushed a fixup. Thanks for the review!

nilsdeppe commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -29,69 +29,71 @@ void weak_divergence(
           } else {
-            auto z_tensor_index = x_tensor_index;
+            const auto z_tensor_index = prepend(result_tensor_index, 2_st);
             z_tensor_index[0] = 2;
```
I failed to remove this line.. I'll do that when I squashed

nilsdeppe commented:
And squashed and updated

fmahebert commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,192 @@
+#include "Utilities/StdArrayHelpers.hpp"
+
+/// \ingroup NumericalAlgorithmsGroup
+/// \brief Compute the weak form divergence of fluxes
```
Document what this means. I'd suggest either adding a formula, or pointing to somewhere else where the formula can be found.

fmahebert commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,192 @@
+template <typename... FluxTags, size_t Dim>
+void weak_divergence(
+    const gsl::not_null<Variables<tmpl::list<Tags::div<FluxTags>...>>*>
+        divergence_of_flux,
```
I suggest `divergence_of_fluxes` (i.e., make plural) to match the argument `fluxes` below

fmahebert commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,192 @@
+    const Variables<tmpl::list<FluxTags...>>& fluxes, const Mesh<Dim>& mesh,
+    const InverseJacobian<DataVector, Dim, Frame::Logical, Frame::Inertial>&
+        det_jac_times_inverse_jacobian) noexcept {
+  if (divergence_of_flux->number_of_grid_points() !=
```
Do you think this warrants an `UNLIKELY` ?

fmahebert commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,192 @@
+          const auto result_tensor_index =
+              result.get_tensor_index(result_storage_index);
+          const auto flux_x_tensor_index = prepend(result_tensor_index, 0_st);
+          const auto flux_y_tensor_index = prepend(result_tensor_index, 1_st);
```
My interpretation is that you'll be using this lambda in 2d, 3d only? To avoid future issues, I would suggest at least documenting the dimension assumptions (add a second line to the comment around line 32?), or perhaps adding a static assert?

nilsdeppe replied:
I've decided to switch the runtime `ERROR` at the end to a `static_assert` and to scope the entire lambda in the not-1d-case

fmahebert commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,192 @@
+        dgemm_<true>(
+            'N', 'N',
+            matrix.rows(),            // rows of matrix and result
+            size / matrix.columns(),  // columns of result and u
```
There is no "u" in this function...

nilsdeppe replied:
But there is a `u` in function :D (I'll fix)

fmahebert commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,192 @@
+        Spectral::weak_flux_differentiation_matrix(mesh.slice_through(0));
+
+    // Compute in eta direction. Since that also needs a transpose, copy into
+    // result, then transpose into `data_buffer`
```
Clarify: compute what in eta direction? Based on the comments below, I suggest "Compute the eta divergence term"

fmahebert commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,192 @@
+    const Matrix& xi_weak_div_matrix =
+        Spectral::weak_flux_differentiation_matrix(mesh.slice_through(0));
+
+    // Compute in zeta direction. Since that also needs a transpose, copy into
```
ditto

fmahebert commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -0,0 +1,192 @@
+    raw_transpose(make_not_null(result_ptr), data_buffer0.data(), chunk_size,
+                  data_buffer1.size() / chunk_size);
+
+    // Compute in eta direction. Since that also needs a transpose, copy into
```
ditto

fmahebert commented on tests/Unit/NumericalAlgorithms/LinearOperators/Test_WeakDivergence.cpp:
```diff
@@ -0,0 +1,163 @@
+    using tag = tmpl::type_from<decltype(tag_v)>;
+    if (mesh.quadrature(0) == Spectral::Quadrature::GaussLobatto) {
+      // Only the interior points are zero. Points on the boundary are non-zero
+      // because they need the flux lifted to be zero.
```
I don't understand what you're saying here (not even enough to ask a specific clarifying question). Can you expand?

Can you explain, in a github response for now, why the result of the divergence should have any dependence on the concept of flux-lifting? This is deeply upsetting if true...

nilsdeppe replied:
Well I think you'll have to be deeply upset then ;)

What happens is you do an integration by-parts, which gives you a boundary contribution and a volume contribution. It is the _sum_ of those that is zero. With GL points the surface integral only contributes to the grid points on the surface, and thus it is only the interior points that are zero. With Gauss points the surface integral contributes to "all" points effectively. Does that help? If not, I can type up some math in a comment, but the short of it is:

P d_i F^i = 0

but for weak form we have

F^i d_i p != P d_i F^i

instead,

P d_i F^i = - F^i d_i P + surface terms = 0

Does that make sense?

fmahebert replied:
👍 Nils and I had a conversation to clear up my misunderstandings .. Nils will add some documentation to clarify what is being tested and why

fmahebert reviewed (changes requested):
Partial review... I stopped when I realized I fundamentally didn't understand what this is doing. Will continue reviewing once my understanding is set straight :)

nilsdeppe commented:
@fmahebert I've pushed a fixup that does what we discussed. Please let me know if I got anything wrong!

fmahebert commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -12,78 +15,76 @@
+ * \ingroup NumericalAlgorithmsGroup
+ * \brief Compute the weak form divergence of fluxes
+ *
+ * We denote the basis funcions by \f$\phi_{\breve{\imath}}\f$. For a
```
1) [trivial, nonbinding] I'd suggest moving this sentence to after the equation. So the structure would be "In a DG scheme we ... For the flux div term this gives ..., where we denote the basis functions by ....". I think this is easier to grab context from :)

2) typo: "funcions" -> "functions" 

3) [nonbinding] drop the breves?

fmahebert commented on src/NumericalAlgorithms/LinearOperators/WeakDivergence.hpp:
```diff
@@ -93,100 +94,152 @@ void weak_divergence(
+                      flux.get(flux_z_tensor_index);
+            }
+          }
+        };
```
[nitpick] Add blank line here

fmahebert commented on tests/Unit/NumericalAlgorithms/LinearOperators/Test_WeakDivergence.cpp:
```diff
@@ -0,0 +1,163 @@
+};
+
+template <size_t Dim>
+void test_weak_divergence_random_jacobian(const Mesh<Dim>& mesh) {
```
So I've thought more about this since we talked yesterday.

I'm concerned that this is not a useful test: by using a constant flux, a non-constant jacobian, and checking the divergence is ~ zero ... is this not functionally the same as the test of the metric-identity jacobian?

Have I missed an important difference?

Perhaps an alternative would be to generate a random flux tensor, compute its divergence using the strong and weak expressions, then test those match at the correct (GL: interior; G: central) points?

nilsdeppe replied:
The reason for the current test is:
- verify that the jacobians are accounted for correctly
- that a consistent differentiation matrix is used in the metric identity jac computation and the div computation
I think checking this before check that the result matches what we expect (e.g. doing what you suggested) is useful because instead of just knowing "something went wrong" it's narrowed down.

I can add the test you suggested. I agree that it's useful

nilsdeppe commented:
@fmahebert I've pushed another fixup commit and rebased on develop

kidder reviewed (commented):
you have my permission to squash

fmahebert commented:
Please add a comment above each of the test functions (there are now two: `test_weak_divergence_random_jacobian` and `test_weak_divergence_constant_jacobian`) saying what is being tested (please use formulas), and clarifying the difference between the two test functions.

Our discussions (and associated documentation improvements) have helped me understand what the _code_ is doing (and, I hope, the clarity in the code for others as well), but I'm still really struggling to understand what the tests are testing and how the work in the tests relates to the math in the WeakDivergence.hpp documentation.

nilsdeppe commented:
Okay, squashed. @fmahebert I added a lengthy paragraph in the `SPECTRE_TEST_CASE` that hopefully explains what and _why_ the tests do what they do. Please let me know how I could improve things!

fmahebert commented on tests/Unit/NumericalAlgorithms/LinearOperators/Test_WeakDivergence.cpp:
```diff
@@ -0,0 +1,365 @@
+  // the interior points are zero. For Gauss points the problem is much harder.
+  // Only the central grid point is zero and only when there is an odd number of
+  // grid points. Verifying that the metric identities are satisfied is a
+  // necessary but not sufficient condition.)
```
Stray ")" ... either add matching "(" somewhere before, or remove.

fmahebert commented on tests/Unit/NumericalAlgorithms/LinearOperators/Test_WeakDivergence.cpp:
```diff
@@ -0,0 +1,365 @@
+    }
+  }
+
+  // Generate constant fluxes that aren't all the same.
```
Is this an out-of-date comment? AFAICT your fluxes here are non-constant (which I think is correct, given that the jacobians are constant)

fmahebert commented on tests/Unit/NumericalAlgorithms/LinearOperators/Test_WeakDivergence.cpp:
```diff
@@ -0,0 +1,365 @@
+  Variables<div_tags> divergence_result{mesh.number_of_grid_points()};
+
+  weak_divergence(make_not_null(&divergence_result), fluxes, mesh,
+                  det_jac_times_inverse_jacobian);
```
[nitpick for legibility] move the blank line on 219 down to below the `weak_divergence` line, to better separate the computation of the weak div from the expected div.

fmahebert reviewed (commented):
YES thank you, this is exactly the kind of explanation of the tests that I was hoping to see. It's super helpful to understand the whole process 👍.

1. Here's three trivial things I saw, please squash them immediately.

2. I do see now what you mean about the "constant jacobian" test being hard... and in the end I'm not even sure how useful of a test it is in the Gauss point case, where there is so much manipulation needed to set up the test. If you wanted to keep only the GaussLobatto case I would not object. But now that you've coded it up and it's working, I'm also happy to see it stay in. Up to you! :)

nilsdeppe commented:
Okay, squashed!

Re 2. I agree it's not a great test, bit messy and requires a lot of gymnastics, but I think now that it's implemented we may as well keep it mostly as a regression test :) For example, if one decides to try to optimize the div function it's good to have the extra check to make sure we don't accidentally break something
